srtcore/common.h

```diff
@@ -167,6 +167,270 @@ const int DEFAULT_LIVE_LATENCY = 120; // (mSec)
+// so that the type is compiler-enforced when calling updateCC() and when
+// connecting signals to slots.
+struct EventVariant
+{
+    enum Type {UNDEFINED, PACKET, ARRAY, ACK, STAGE, INIT} type;
+    union U
+    {
+        CPacket* packet;
+        uint32_t ack;
+        struct
+        {
+            int32_t* ptr;
+            size_t len;
+        } array;
+        ECheckTimerStage stage;
+        EInitEvent init;
+    } u;
+
+    EventVariant()
+    {
+        type = UNDEFINED;
+        memset(&u, 0, sizeof u);
+    }
+
+    template<Type t>
+    struct VariantFor;
+
+    template <Type tp, typename Arg>
+    void Assign(Arg arg)
+    {
+        type = tp;
+        (u.*(VariantFor<tp>::field())) = arg;
+        //(u.*field) = arg;
+    }
+
+    void operator=(CPacket* arg) { Assign<PACKET>(arg); };
+    void operator=(uint32_t arg) { Assign<ACK>(arg); };
+    void operator=(ECheckTimerStage arg) { Assign<STAGE>(arg); };
+    void operator=(EInitEvent arg) { Assign<INIT>(arg); };
+
+    // Note: UNDEFINED and ARRAY don't have assignment operator.
+    // For ARRAY you'll use 'set' function. For UNDEFINED there's nothing.
+
+
+    template <class T>
+    EventVariant(T arg)
+    {
+        *this = arg;
+    }
+
+    int32_t* get_ptr()
+    {
+        return u.array.ptr;
+    }
+
+    size_t get_len()
+    {
+        return u.array.len;
+    }
+
+    void set(int32_t* ptr, size_t len)
+    {
+        type = ARRAY;
+        u.array.ptr = ptr;
+        u.array.len = len;
+    }
+
+    EventVariant(int32_t* ptr, size_t len)
+    {
+        set(ptr, len);
+    }
+
+    template<Type T>
+    typename VariantFor<T>::type get()
+    {
+        return u.*(VariantFor<T>::field());
+    }
+};
+
+/*
+    Maybe later.
+    This had to be a solution for automatic extraction of the
+    type hidden in particular EventArg for particular event so
+    that it's not runtime-mistaken.
+
+    In order that this make sense there would be required an array
+    indexed by event id (just like a slot array m_Slots in CUDT),
+    where the "type distiller" function would be extracted and then
+    combined with the user-connected slot function this would call
+    it already with correct type. Note that also the ConnectSignal
+    function would have to get the signal id by template parameter,
+    not function parameter. For example:
+
+    m_parent->ConnectSignal<TEV_ACK>(SSLOT(updateOnSent));
+
+    in which updateOnSent would have to receive an appropriate type.
+    This has a disadvantage that you can't connect multiple signals
+    with different argument types to the same slot, you'd have to
+    make slot wrappers to translate arguments.
+
+    It seems that a better idea would be to create binders that would
+    translate the argument from EventArg to the correct type according
+    to the rules imposed by particular event id. But I'd not make it
+    until there's a green light on C++11 for SRT, so maybe in a far future.
```

Comment:
Boost actually would be an option, didn't think about this earlier, might be a nice idea. But this needs some more discussion about that; it would actually save us a lot of trouble with "handworking".